maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java

```
@@ -571,25 +587,26 @@ private void uploadChecksums( File file, byte[] bytes )
             }
             try
             {
-                Set<String> algos = new HashSet<>();
-                for ( RepositoryLayout.Checksum checksum : checksums )
+                ArrayList<ChecksumAlgorithmFactory> algorithms = new ArrayList<>();
+                for ( RepositoryLayout.ChecksumLocation checksumLocation : checksums )
                 {
-                    algos.add( checksum.getAlgorithm() );
+                    algorithms.add( checksumLocation.getChecksumAlgorithmFactory() );
                 }
 
-                Map<String, Object> sumsByAlgo;
+                Map<String, String> sumsByAlgo;
```

Since the value is not an object anymore `uploadChecksum()` cannot throw with `checksum instanceof Exception`. Can you explain how this is supposed to work?

---

The `sumsByAlgo` was always populated from (now deprecated) `ChecksumUtils.calc` method, that in turn returned non-string in map value ONLY in case of `NoSuchAlgorithmException`. This PR OTOH makes sure that checksums "fail fast", so you cannot even start (get factory from selector) for algorithm that does not exists. So to say, old code was more forgiving if you asked for algorithm not supported by MessageDigester (all this as actual "name evaluation" was happening way before, and it just accepted any string), while in this PR you must pass in a factory, that you can obtain ONLY for existing algorithms.